core/src/main/java/jenkins/scm/RunWithSCMMixIn.java

```diff
@@ -0,0 +1,161 @@
+                if (pr != null && pr.isWorseThan(Result.SUCCESS)) {
+                    // we are still building, so this is just the current latest information,
+                    // but we seems to be failing so far, so inherit culprits from the previous build.
+                    // isBuilding() check is to avoid recursion when loading data from old Hudson, which doesn't record
+                    // this information
+                    r.addAll(p.getCulprits());
+                }
+            }
+            for (ChangeLogSet<? extends ChangeLogSet.Entry> c : getChangeSets()) {
+                for (ChangeLogSet.Entry e : c)
+                    r.add(e.getAuthor());
+            }
+
+            if (p instanceof AbstractBuild && upstreamCulprits) {
+                // If we have dependencies since the last successful build, add their authors to our list
+                if (p.getPreviousNotFailedBuild() != null) {
+                    Map<AbstractProject, AbstractBuild.DependencyChange> depmap =
+                            ((AbstractBuild<?,?>) p).getDependencyChanges((AbstractBuild<?,?>)p.getPreviousSuccessfulBuild());
+                    for (AbstractBuild.DependencyChange dep : depmap.values()) {
+                        for (AbstractBuild<?, ?> b : dep.getBuilds()) {
+                            for (ChangeLogSet.Entry entry : b.getChangeSet()) {
+                                r.add(entry.getAuthor());
+                            }
+                        }
+                    }
+                }
+            }
+
+            return r;
+        }
+
+        return new AbstractSet<User>() {
+            public Iterator<User> iterator() {
+                return new AdaptedIterator<String,User>(asRun().getCulpritIds().iterator()) {
+                    protected User adapt(String id) {
+                        return User.get(id);
+                    }
+                };
+            }
+
+            public int size() {
+                return asRun().getCulpritIds().size();
+            }
+        };
+    }
+
+    /**
+     * Returns true if this user has made a commit to this build.
+     */
+    public boolean hasParticipant(User user) {
+        for (ChangeLogSet<? extends ChangeLogSet.Entry> c : getChangeSets()) {
+            for (ChangeLogSet.Entry e : c)
+                try {
+                    if (e.getAuthor() == user)
+                        return true;
+                } catch (RuntimeException re) {
+                    LOGGER.log(Level.INFO, "Failed to determine author of changelog " + e.getCommitId() + "for " + asRun().getParent().getDisplayName() + ", " + asRun().getDisplayName(), re);
+                }
+        }
+        return false;
+    }
+
+    public interface RunWithSCM<JobT extends Job<JobT, RunT> & Queue.Task,
+            RunT extends Run<JobT, RunT> & RunWithSCM<JobT,RunT> & Queue.Executable> {
+        @Nonnull
+        List<ChangeLogSet<? extends ChangeLogSet.Entry>> getChangeSets();
```

Comment:
Well the principle seems to work in #2864, though it is messier there since

* this was existing code designed to work in `-source 7` (or maybe it was even `-source 6` at the time, I cannot recall)
* a lot of the methods were defined in other types

Here I think it should be cleaner. Delete `RunWithSCM`, `asRun`, and `getRunWithSCMMixIn`, and make `RunWithSCMMixIn` into an `interface`, with `default` on as many methods as possible.